c++/src/Reader.cc

```diff
@@ -1364,6 +1365,111 @@ namespace orc {
     int64_t getEpochOffset() const override;
   };
 
+  uint64_t maxStreamsForType(const proto::Type& type) {
+    switch (static_cast<int64_t>(type.kind())) {
+      case proto::Type_Kind_STRUCT:
+        return 1;
+      case proto::Type_Kind_INT:
+      case proto::Type_Kind_LONG:
+      case proto::Type_Kind_SHORT:
+      case proto::Type_Kind_FLOAT:
+      case proto::Type_Kind_DOUBLE:
+      case proto::Type_Kind_BOOLEAN:
+      case proto::Type_Kind_BYTE:
+      case proto::Type_Kind_DATE:
+      case proto::Type_Kind_LIST:
+      case proto::Type_Kind_MAP:
+      case proto::Type_Kind_UNION:
+        return 2;
+      case proto::Type_Kind_BINARY:
+      case proto::Type_Kind_DECIMAL:
+      case proto::Type_Kind_TIMESTAMP:
+        return 3;
+      case proto::Type_Kind_CHAR:
+      case proto::Type_Kind_STRING:
+      case proto::Type_Kind_VARCHAR:
+        return 4;
+      default:
+          return 0;
+      }
+  }
+
+  uint64_t ReaderImpl::getMemoryUse(int stripeIx) {
+    uint64_t maxDataLength = 0;
+
+    if (stripeIx >= 0 && stripeIx < footer->stripes_size()) {
+      uint64_t stripe = footer->stripes(stripeIx).datalength();
+      if (maxDataLength < stripe) {
+        maxDataLength = stripe;
+      }
+    } else {
+      for (int i=0; i < footer->stripes_size(); i++) {
+        uint64_t stripe = footer->stripes(i).datalength();
+        if (maxDataLength < stripe) {
+          maxDataLength = stripe;
+        }
+      }
+    }
+
+    bool hasStringColumn = false;
+    uint64_t nSelectedStreams = 0;
+    for (int i=0; !hasStringColumn && i < footer->types_size(); i++) {
+      if (selectedColumns[i]) {
```

Comment:
clang needs:
if (selectedColumns[static_cast<size_t>(i)]) {

to avoid a message about using signed int where an unsigned int is needed.